sound/soc/sof/intel/hda.c

```diff
@@ -1118,7 +1119,7 @@ static int hda_sdw_machine_select(struct snd_sof_dev *sdev)
 					break;
 			}
 			/* Found if all Slaves are checked */
-			if (i == hdev->info.count)
+			if (!link->num_adr)
```

Comment:
shouldn't this be `if (i == hdev->info.count || !link->num_adr)`? Because if you terminate the loop on the `i < hdev->info.count` condition, `link` will point to an invalid entry?

Reply:
at this time link will point to an empty entry which indicates the end of link, so the matching flow is complete.  Please check the following cases.
````C
static const struct snd_soc_acpi_link_adr tgl_i2s_rt1308[] = {
	{
		.mask = BIT(0),
		.num_adr = ARRAY_SIZE(rt711_0_adr),
		.adr_d = rt711_0_adr,
	},
	{}
};

static const struct snd_soc_acpi_link_adr tgl_rvp[] = {
	{
		.mask = BIT(0),
		.num_adr = ARRAY_SIZE(rt711_0_adr),
		.adr_d = rt711_0_adr,
	},
	{
		.mask = BIT(1),
		.num_adr = ARRAY_SIZE(rt1308_1_adr),
		.adr_d = rt1308_1_adr,
	},
	{}
};
````

Reply:
@RanderWang not really. hdev->info.count comes from the parsing of ACPI table, something could be wrong. Whereas the link tables are constant. Let's add the test and be consistent. It may be overkill in most cases but will guarantee we deal with errors in the same way inside the loop and outside. Thanks!

Reply:
yes, defensive programming.